Mathlib/Algebra/Hom/Centroid.lean

```diff
@@ -0,0 +1,532 @@
+instance hasNsmul : SMul ℕ (CentroidHom α) :=
+  ⟨fun n f ↦
+    {
+      ((SMul.smul n f) : α →+ α) with
+        map_mul_left' := fun a b ↦ by
+          change n • f (a * b) = a * n • f b
+          rw [map_mul_left f, ← mul_smul_comm]
+        map_mul_right' := fun a b ↦ by
+          change n • f (a * b) = n • f a * b
+          rw [map_mul_right f, ← smul_mul_assoc] }⟩
+#align centroid_hom.has_nsmul CentroidHom.hasNsmul
+
+instance hasNpowNat : Pow (CentroidHom α) ℕ :=
+  ⟨fun f n ↦
+    {
+      (f.toEnd ^ n :
+        AddMonoid.End
+          α) with
+      map_mul_left' := fun a b ↦ by
+        induction' n with n ih
+        · exact rfl
+        · simp
+          rw [pow_succ]
+          exact (congr_arg f.toEnd ih).trans (f.map_mul_left' _ _)
+      map_mul_right' := fun a b ↦ by
+        induction' n with n ih
+        · exact rfl
+        · simp
+          rw [pow_succ]
+          exact (congr_arg f.toEnd ih).trans (f.map_mul_right' _ _) }⟩
+#align centroid_hom.has_npow_nat CentroidHom.hasNpowNat
+
+@[simp, norm_cast]
+theorem coe_zero : ⇑(0 : CentroidHom α) = 0 :=
+  rfl
+#align centroid_hom.coe_zero CentroidHom.coe_zero
+
+@[simp, norm_cast]
+theorem coe_one : ⇑(1 : CentroidHom α) = id :=
+  rfl
+#align centroid_hom.coe_one CentroidHom.coe_one
+
+@[simp, norm_cast]
+theorem coe_add (f g : CentroidHom α) : ⇑(f + g) = f + g :=
+  rfl
+#align centroid_hom.coe_add CentroidHom.coe_add
+
+@[simp, norm_cast]
+theorem coe_mul (f g : CentroidHom α) : ⇑(f * g) = f ∘ g :=
+  rfl
+#align centroid_hom.coe_mul CentroidHom.coe_mul
+
+-- Eligible for `dsimp`
+@[simp, norm_cast, nolint simpNF]
+theorem coe_nsmul (f : CentroidHom α) (n : ℕ) : ⇑(n • f) = n • (⇑f) :=
+  rfl
+#align centroid_hom.coe_nsmul CentroidHom.coe_nsmul
+
+@[simp]
+theorem zero_apply (a : α) : (0 : CentroidHom α) a = 0 :=
+  rfl
+#align centroid_hom.zero_apply CentroidHom.zero_apply
+
+@[simp]
+theorem one_apply (a : α) : (1 : CentroidHom α) a = a :=
+  rfl
+#align centroid_hom.one_apply CentroidHom.one_apply
+
+@[simp]
+theorem add_apply (f g : CentroidHom α) (a : α) : (f + g) a = f a + g a :=
+  rfl
+#align centroid_hom.add_apply CentroidHom.add_apply
+
+@[simp]
+theorem mul_apply (f g : CentroidHom α) (a : α) : (f * g) a = f (g a) :=
+  rfl
+#align centroid_hom.mul_apply CentroidHom.mul_apply
+
+-- Eligible for `dsimp`
+@[simp, nolint simpNF]
+theorem nsmul_apply (f : CentroidHom α) (n : ℕ) (a : α) : (n • f) a = n • f a :=
+  rfl
+#align centroid_hom.nsmul_apply CentroidHom.nsmul_apply
+
+@[simp]
+theorem to_End_zero : (0 : CentroidHom α).toEnd = 0 :=
+  rfl
+#align centroid_hom.to_End_zero CentroidHom.to_End_zero
+
+@[simp]
+theorem to_End_add (x y : CentroidHom α) : (x + y).toEnd = x.toEnd + y.toEnd :=
+  rfl
+#align centroid_hom.to_End_add CentroidHom.to_End_add
+
+theorem to_End_nsmul (x : CentroidHom α) (n : ℕ) : (n • x).toEnd = n • x.toEnd :=
+  rfl
+#align centroid_hom.to_End_nsmul CentroidHom.to_End_nsmul
+
+-- Porting note: I guess the porter has naming issues still
```

Comment:
What does this mean?

Reply:
I phrased it poorly, but I was referring to the fact that I had to change `coe_to_add_monoid_hom_injective.AddCommMonoid` to `coe_to_add_monoid_hom_injective.addCommMonoid`